Sources/TuistAutomation/ProjectMappers/PruneUITestsProjectMapper.swift

```diff
@@ -0,0 +1,21 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistGraph
+
+public final class PruneUITestsProjectMapper: ProjectMapping {
+    public init() {}
+
+    public func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
```

Comment:
I think we should revisit the idea of "pruning". I think long-term, it's better if mappers just do what they are designed for. In this case removing certain targets for the project. This might leave the project in a invalid state, for example schemes with broken references to targets. For that reason I think we should extract from `CacheTreeShakingGraphMapper` the logic that ensures the integrity of the project. We can place that in a `IntegrityGraphMapper`. @luispadron do you think "integrity" is the right word for describe what it does?

Reply:
I like the idea, however maybe we can tackle that in the following PR from two reasons:
* I'd like to keep the size of a pull request small 
* getting rid of puring feels out of scope for this PR 

I have already been playing with extracting the integrity part from `CacheTreeShakingGraphMapper`. I'd like to write down a quick summary to be sure we are on the same page 🙏  

1. `SkipUITestsProjectMapper` would remove the ui tests targets explicitly instead of marking them as to be pruned. 
```swift
public final class SkipUITestsProjectMapper: ProjectMapping {
    public func map(project: Project) throws -> (Project, [SideEffectDescriptor]) {
        var project = project
        project.targets = project.targets.filter { $0.product != .uiTests }
        return (project, [])
    }
}
```

2. `IntegrityGraphMapper` implementation would be identical to the current state of `CacheTreeShakingGraphMapper`, but it would do nothing related to `target.prune`, meaning this would be removed: 
https://github.com/tuist/tuist/blob/40982d094e61247c263670d86c0140e77eb76619/Sources/TuistCache/GraphMappers/CacheTreeShakingGraphMapper.swift#L13

3. We'd need to add `IntegrityGraphMapper` as a last mapper to every place where the `CacheTreeShakingGraphMapper` is added. 

4. `CacheGraphMutator` sets targets to be pruned - how we would migrate this part? Should the mutator remove the targets directly? Would we need `CacheTreeShakingGraphMapper` then still? I am a bit confused how those would fit together. 

Thanks for the feedback!

Reply:
Hope you do not mind, but I have created a follow-up ticket to tackle pruning refactoring 🙇 

* https://github.com/tuist/tuist/issues/3078